File: chess/pieces/pawn.py
from chess import PieceColorEnum
from chess.pieces.default_piece import DefaultPiece
from chess.pieces.piece import BoardSize, Piece, PiecePosition


class Pawn(DefaultPiece):
    __vertical_direction_multiplier = 1

    def __init__(self, color: PieceColorEnum, board_size: BoardSize, position: PiecePosition) -> None:
        if color == PieceColorEnum.WHITE:
            self.__vertical_direction_multiplier = -1

        super().__init__(color, board_size, position)

    def update_movements(self, board: list[list[Piece | None]]) -> None:
        self.clean_movements()
        y = self.position['y'] + 1 * self.__vertical_direction_multiplier

        if 0 <= y < self.board_size['height'] and board[y][self.position['x']] == None:
            self.movements[y][self.position['x']] = True
        

File: src/main/java/com/vini/app/pieces/Pawn.java
package com.vini.app.pieces;

import java.util.List;
import java.util.Objects;

import com.vini.app.types.ColorEnum;

public class Pawn extends Piece {
	private final int[][] forkDirections = {{-1, 1}, {1, 1}};
	private Boolean hasMoved = false;
	private int directionWeight = 1;

	@Override
	public void move(List<List<IPiece>> board, int[] position) {
		super.move(board, position);

		if (!this.hasMoved) {
			this.hasMoved = true;
		}
	}

	@Override
	public IPiece updateMoves(final List<List<IPiece>> board) {
		int maxDistance = 1;

		if (!this.hasMoved) {
			maxDistance = 2;
		}

		for (int i = 1; i <= maxDistance; i++) {
			int[] targetPosition = {
				this.position()[0],
				this.position()[1] + i * this.directionWeight
			};

			if (!this.insideBoard(board, targetPosition)) {
				break;
			};

			IPiece target = board.get(targetPosition[1]).get(targetPosition[0]);

			if (Objects.nonNull(target)) {
				break;
			}
			this.moves().get(targetPosition[1]).set(targetPosition[0], true);
		}

		for (int[] forkDirection : this.forkDirections) {
			int[] targetPosition = {
				this.position()[0] + forkDirection[0],
				this.position()[1] + forkDirection[1] * this.directionWeight
			};

			if (!this.insideBoard(board, targetPosition)) {
				continue;
			}

			IPiece target = board.get(targetPosition[1]).get(targetPosition[0]);

			if (Objects.isNull(target) || this.isFriend(this, target)) {
				continue;	
			}

			this.moves().get(targetPosition[1]).set(targetPosition[0], true);
		}

		return this;
	}

	public IPiece setColor(ColorEnum color) {
		super.setColor(color);
		if (color == ColorEnum.WHITE) {
			this.directionWeight = -1;
		}
		return this;
	}
}

